### PR TITLE
13851 Added improvements for credit memo total fields

### LINKED ISSUE #13851: Credit doesn't recognize amount after comma
When I enter an amount with decimals (like 2,50) in the field "Adjustment Refund" while making a Credit Memo the amount on the real credit memo is without decimals (2,00). So the Credit Memo isn't in line with the real credit amount.

### Preconditions
1. Magento 2.1.10

### Steps to reproduce
1. Create a Credit Memo for an order
2. Enter an amount with decimals (by example 1,23) in the field Adjustment Refund
3. Create an offline payment (finish credit memo)

### Expected result
1. The credit memo should show the amount of 1,23 in the field "Adjustment Refund" on the Credit Memo overview

### Actual result
1. The credit memo should show the amount of 1,00 in the field "Adjustment Refund" on the Credit Memo overview

I'm using a comma as decimal seperator, because my local language is dutch.

### PR DESCRIPTION
**Pull request by sergiy-v**
### Description
Added improvements for credit memo (new page) total fields.

### Related Pull Requests

#27492

### Fixed Issues
#13851

### Manual testing scenarios
See #13851 issue description for details.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sergiy-v** [comment]
work in progress

File: app/code/Magento/Sales/Model/Order/CreditmemoFactory.php [hunk 1]
```diff
@@ -41,15 +47,18 @@ class CreditmemoFactory
...
     public function __construct(
         \Magento\Sales\Model\Convert\OrderFactory $convertOrderFactory,
         \Magento\Tax\Model\Config $taxConfig,
+        FormatInterface $localeFormat,
         \Magento\Framework\Serialize\Serializer\Json $serializer = null
     ) {
         $this->convertor = $convertOrderFactory->create();
         $this->taxConfig = $taxConfig;
+        $this->localeFormat = $localeFormat;
         $this->serializer = $serializer ?: \Magento\Framework\App\ObjectManager::getInstance()->get(
```
Comment (lbajsarowicz) [inline app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: Remove backward compatible class loader and add your dependency as last.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please keep backwards compatible constructors :-)

File: app/code/Magento/Sales/Model/Order/CreditmemoFactory.php [hunk 2]
```diff
@@ -42,17 +49,20 @@ class CreditmemoFactory
...
         \Magento\Tax\Model\Config $taxConfig,
-        \Magento\Framework\Serialize\Serializer\Json $serializer = null
+        \Magento\Framework\Serialize\Serializer\Json $serializer = null,
+        FormatInterface $localeFormat = null
     ) {
         $this->convertor = $convertOrderFactory->create();
         $this->taxConfig = $taxConfig;
-        $this->serializer = $serializer ?: \Magento\Framework\App\ObjectManager::getInstance()->get(
+        $this->serializer = $serializer ?: ObjectManager::getInstance()->get(
             \Magento\Framework\Serialize\Serializer\Json::class
```
Comment (lbajsarowicz) [inline app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: You can also import that class. I recommend you to use alias `JsonSerializer` to be explicit what it does.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Only import the `Json` class and fix the Static Tests failure.

File: app/code/Magento/Sales/Model/Order/CreditmemoFactory.php (hunk 1 above)
Comment (sergiy-v) [reply app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: The changes has been added, thank you.

File: app/code/Magento/Sales/Model/Order/CreditmemoFactory.php (hunk 2 above)
Comment (sergiy-v) [reply app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: The changes has been added, thank you for recommendation.

**Comment by sergiy-v** [comment]
I believe the tests errors not related to the changes, @lbajsarowicz could you please check it.

Thank you.

**Comment by lbajsarowicz** [comment]
@magento run all tests
Let's restart them.

**Comment by sergiy-v** [comment]
@lbajsarowicz please check the changes once you will have a chance.

Thank you.

**Review by eduard13 — APPROVED** [review]
The failing tests aren't related with this PR. I've linked the PR that should fix the failing Unit Test.
Thank you.

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-7223](https://jira.corp.magento.com/browse/ENGCOM-7223) has been created to process this Pull Request

**Comment by eduard13** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**
1. Create a Credit Memo for an order;
2. Enter an amount with decimals in the field Adjustment Refund;
3. Create an offline payment (finish credit memo)

**Before:**
creditmemobefore

**After:**
credit after

**Comment by slavvka** [comment]
@magento run all tests

**Comment by sdzhepa** [comment]
@engcom-Alfa 

After reviewing with Product managers was decided to add the PR to the next `2.4.0` Magento release. 
But this functionality affect's  order management process and financial documents
So it was decided to do extended testing for this PR additionally to be sure that we do not brake or change any functionality around it.

I would request to check the next scenarios:
- full credit memo
- partial credit memo
- cases with multiple currencies configured on the store(e.g. default/base and admin but different on a couple of websites/store-views)
- Localization settings, like digit delimiter, is "," or "."  (12,01 and 12.01)
- reports about the credit memo shown in the email (numbers are correct?)
- reports about credit memo shown on Storefront(My account) as Comments related to order
- advanced ability "Notify Customer by email" from Admin Panel
- Orders with taxes
- Orders with discounts/reward points/store credits etc
- ...
- any additional user flows and cases that I could miss in this list


As a QA report should be:
- Short description of what was tested
- If need more details about precondition or additional configuration
- Confirmation that bugs if fixed and/or other functionality works as previously(expected)

**Comment by sdzhepa** [comment]
Hello @engcom-Alfa 

I know that you have tested done additional and advanced testing for this PR before the merge.
We discussed it in today internal sync.
Could you please add here a detailed QA report(what was tested) to be transparent here?

**Comment by engcom-Alfa** [comment]
Hello @sdzhepa 

I checked the PR and everything works well.
Fixed the issue described in #13851 and other functionality works as previously (expected).

### _What was tested:_

**1. The issue described in #13851** 

- Create a Credit Memo for an order

- Enter an amount with decimals (by example 1.23) in the fields Adjustment Refund, Adjustment Fee

- Create an offline payment (finish credit memo)

**Before:** :heavy_multiplication_x: 




**After:** :heavy_check_mark: 




**2. Also checked additional cases:**

- [ :heavy_check_mark: ] **full credit memo**;

screenshot1

- [ :heavy_check_mark: ] **partial credit memo**;

screenshot2

- [ :heavy_check_mark: ]  **credit memo on other locale (_German_ locale in my case) with digit delimiter ","**;





- [ :heavy_check_mark: ] **multiple currencies configured on the store (in my case $ and €)**;

screenshot3

- [ :heavy_check_mark: ] **credit memo with taxes (including and excluding)**;

Screenshot from 2020-04-10 22-43-50

screenshot4

- [ :heavy_check_mark: ] **credit memo on Storefront(My account) with related to order comments**

screenshot6

- [ :heavy_check_mark: ]  :ballot_box_with_check: **Email Copy of Credit Memo**

screenshot7
screenshot8

- [ :heavy_check_mark: ] **Print PDF credit memo**

Screenshot from 2020-04-10 22-58-08

**Pull request merged**